cpp/oneapi/dal/test/communicator.cpp

```diff
@@ -0,0 +1,283 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <mutex>
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/detail/communicator.hpp"
+#include "oneapi/dal/test/engine/tables.hpp"
+#include "oneapi/dal/test/engine/thread_communicator.hpp"
+
+namespace oneapi::dal::test {
+
+namespace te = dal::test::engine;
+
+class communicator_test {
+public:
+    dal::detail::spmd_communicator create_communicator(std::int64_t rank_count) {
+        CAPTURE(rank_count);
+        thread_comm_.reset(new te::thread_communicator{ rank_count });
+        return dal::detail::spmd_communicator{ *thread_comm_ };
+    }
+
+    dal::detail::spmd_communicator create_communicator() {
+        return create_communicator(generate_thread_count());
+    }
+
+    std::int64_t generate_thread_count() const {
+        return GENERATE(1, 2, 4, 8);
+    }
+
+    template <typename Body>
+    void execute(Body&& body) {
+        if (!thread_comm_) {
+            create_communicator(generate_thread_count());
+        }
+        thread_comm_->execute(body);
+    }
+
+    template <typename Body>
+    void exclusive(Body&& body) {
+        std::scoped_lock<std::mutex> lock{ mtx_ };
+        body();
+    }
+
+    table create_random_table(std::int64_t row_count, std::int64_t column_count) {
+        std::uniform_real_distribution<float> uniform{ -10.f, 10.f };
+
+        auto x = array<float>::empty(row_count * column_count);
+        float* x_ptr = x.get_mutable_data();
+
+        for (std::int64_t i = 0; i < x.get_count(); i++) {
+            x_ptr[i] = uniform(rng_);
+        }
+
+        return homogen_table::wrap(x, row_count, column_count);
+    }
+
+    std::vector<table> create_random_tables(std::int64_t count) {
+        std::uniform_int_distribution<std::int64_t> uniform{ 10, 30 };
+
+        std::vector<table> tables;
+        tables.reserve(count);
+
+        for (std::int64_t i = 0; i < count; i++) {
+            const std::int64_t row_count = uniform(rng_);
+            const std::int64_t column_count = uniform(rng_);
+            tables.push_back(create_random_table(row_count, column_count));
+        }
+
+        return tables;
+    }
+
+private:
+    static constexpr int seed = 7777;
```

Comment:
There are 777, 7777 and 777777 seeds ...

Reply:
What do you suggest? It could be separate refactoring task of replacing all seeds to predefined constant.